### PR TITLE
Add model and register versions

### LINKED ISSUE #203: Register models in the example repos to see them in the MR
Feedback by @omesser in one of the Studio tickets.

### PR DESCRIPTION
**Pull request by aguschin**
close #203

### CODE REVIEW

File: example-get-started/generate.sh [hunk 1]
```diff
@@ -135,6 +143,20 @@ dvc stage add -n train \
   -o model.pkl \
   python src/train.py data/features model.pkl
 dvc repro
+python <<EOF
+from dvc.repo import Repo
```
Comment (shcheklein) [inline example-get-started/generate.sh:147]: feel we need a command for that? :)

may be it's easier append to dvc.yaml via bash?

Comment (aguschin) [reply example-get-started/generate.sh:147]: yes, we need that command I guess!

Let's use python approach if no strong objections.

Comment (shcheklein) [reply example-get-started/generate.sh:147]: yep, no strong objections at all. Can use python, no worries. Thanks @aguschin for updating this repo.

File: example-get-started/code/src/featurization.py [hunk 2]
```diff
@@ -56,7 +56,7 @@ def generate_and_save_train_features(train_input, train_output, bag_of_words, tf
         tfidf (sklearn.feature_extraction.text.TfidfTransformer): TF-IDF transformer.
     """
     df_train = get_df(train_input)
-    train_words = np.array(df_train.text.str.lower().values.astype("U"))
+    train_words = np.array(df_train.text.str.lower().values)
```
Comment (shcheklein) [inline example-get-started/code/src/featurization.py:59]: curious - why this change?

**Review by shcheklein — APPROVED** [review]
just one comment, otherwise should be good to go?

File: example-get-started/generate.sh [hunk 3]
```diff
@@ -135,6 +143,20 @@ dvc stage add -n train \
...
+repo = Repo(".")
+artifact = Artifact(
+  path="model.pkl", 
+  type="model",
+  desc="Detect whether the given stackoverflow question should have R language tag",
+  labels=["nlp", "classification", "stackoverflow"]
+)
+repo.artifacts.add("text-classification", artifact)
+EOF
+
```
Comment (dberenbaum) [inline example-get-started/generate.sh:159]: Why not change the source code to use `live.log_artifact()`?

File: example-get-started/code/src/featurization.py (hunk 2 above)
Comment (aguschin) [reply example-get-started/code/src/featurization.py:59]: It was breaking (just terminating the process without any hints). Not sure what was the problem. And IIUC, `.astype("U")` convert value to string, but it's already string. So not sure why it was needed in the first place :)

File: example-get-started/generate.sh (hunk 3 above)
Comment (aguschin) [reply example-get-started/generate.sh:159]: Right 😂 Not sure how I missed that. Will do!

File: example-get-started/code/src/featurization.py (hunk 2 above)
Comment (shcheklein) [reply example-get-started/code/src/featurization.py:59]: @aguschin can we check blame please and see who put it there in the first place and ask for the review? (can be me of course :) just in case if @daavoo and @dberenbaum remember something)

Comment (aguschin) [reply example-get-started/code/src/featurization.py:59]: This goes all the way back to the creation of this repo 5 years ago: https://github.com/iterative/example-repos-dev/blame/e22378ce6a7c1729d3d5bdfe365be38bad605353/code/featurization.py

it could be python 2 (for which this operation made sense IIUC)

don't think it's relevant to double-check this :)

**Comment by aguschin** [comment]
For the record, dvclive's `log_artifact` doesn't add artifact to `dvc.yaml` right now. @daavoo is looking into it.

File: example-get-started-experiments/README.md [hunk 4]
```diff
@@ -52,7 +52,14 @@ and create a P.R. from the best of the 3 experiments that are found in the lates
 
 - Add a model to Studio Model Registry
 
-Go to Studio MR and click "Add a model". Fill the `path=models/model.pkl` and add a model to a separate branch. 
+Go to Studio MR and click "Add a model". Details:
```
Comment (shcheklein) [inline example-get-started-experiments/README.md:55]: it's not related?, let's do a separate PR please

File: example-get-started/generate.sh [hunk 5]
```diff
@@ -63,7 +63,16 @@ git tag -a "1-dvc-init" -m "DVC initialized."
 mkdir data
 dvc get https://github.com/iterative/dataset-registry \
   get-started/data.xml -o data/data.xml
-dvc add data/data.xml --desc "Initial XML StackOverflow dataset (raw data)"
+python <<EOF
+from dvc.repo import Repo
```
Comment (shcheklein) [inline example-get-started/generate.sh:67]: can we also use `log_artifact`?

File: example-get-started/generate.sh [hunk 6]
```diff
@@ -170,9 +178,11 @@ echo "plots:
 dvc repro
 git add .gitignore dvc.yaml dvc.lock eval
 tick
-git commit -m "Create evaluation stage"
+git commit -am "Create evaluation stage"
 git tag -a "8-evaluation" -m "Baseline evaluation stage created."
 git tag -a "baseline-experiment" -m "Baseline experiment evaluation"
+gto register dvclive:text-classification --version v1.0.0
```
Comment (shcheklein) [inline example-get-started/generate.sh:184]: do we use dvclive with a custom path in the eval stage? let's try to reuse it ... I don't think we want to have the `dvclive` dir in this project ... or at least is should be a single dir with everything in it

**Comment by daavoo** [comment]
I am ok with going with the P.R. (after addressing the comments) but it feels to me that we need to take a step back and discuss the "Model Registry using DVCLive log_artifact" workflow.

I didn't play with it until now because there was no Studio support, but I have now tried to set up an example myself (and reviewed this P.R.) and the experience feels unpolished.

File: example-get-started/generate.sh (hunk 5 above)
Comment (aguschin) [reply example-get-started/generate.sh:67]: We can, but I don't think it's logical (it's either a separate python script or modifying `prepare.py`, which doesn't make sense to me). Rather I can simply echo few lines from bash if that's better.

File: example-get-started/generate.sh (hunk 6 above)
Comment (aguschin) [reply example-get-started/generate.sh:184]: Looks somewhat strange to me to have a model registered in `eval/live` TBH. For this particular case, it be good to write to `dvc.yaml` in the root, e.g. using
```
with Live("") as live:
  ...
```
Wdyt @daavoo?

File: example-get-started/generate.sh (hunk 5 above)
Comment (aguschin) [reply example-get-started/generate.sh:67]: wdyt?

File: example-get-started/generate.sh (hunk 6 above)
Comment (daavoo) [reply example-get-started/generate.sh:184]: > it be good to write to dvc.yaml in the root, e.g. using

I think this would not work or would work in unexpected ways.

We could update the repo to have a parent  `results` output folder and use `results/train` and `results/eval` in each stage, I think it is what Ivan was suggesting

Comment (aguschin) [reply example-get-started/generate.sh:184]: Looks like a potentially substantial update to me (I assume we should move `model.pkl` to `results/train` as well?). I suggest to keep this repo simple. Let's use the approach with `results/...` for the `example-gs-experiments` repo and here simply edit `dvc.yaml` in the root like before by calling DVC API.

Comment (aguschin) [reply example-get-started/generate.sh:184]: so unless you're against, I'm reverting the changes.

**Comment by aguschin** [comment]
This is how it looks like now:
image

**Comment by aguschin** [comment]
ok, I re-generated both repos and pushed changes. In Studio everything looks good AFAIT.

@daavoo, can I ask you to review changes for example-get-started-experiments? Need a second look in case I missed something. Thanks!

File: example-get-started-experiments/code/src/train.py [hunk 7]
```diff
@@ -55,17 +56,25 @@ def train():
...
-    learn.fine_tune(
-        **params.train.fine_tune_args,
-        cbs=[DVCLiveCallback(dir="results/train", report="md")],
-    )
-    models_dir = Path("models")
-    models_dir.mkdir(exist_ok=True)
-    learn.export(fname=(models_dir / "model.pkl").absolute())
+        learn.fine_tune(
+            **params.train.fine_tune_args,
+            cbs=[DVCLiveCallback(dir="results/train", report="md")],
```
Comment (daavoo) [inline example-get-started-experiments/code/src/train.py:66]:
```suggestion
            cbs=[DVCLiveCallback(live=live)],
```
https://dvc.org/doc/dvclive/ml-frameworks/fastai#examples
